### ISSUE TITLE
Post picture from clipboard

### ISSUE THREAD

**Issue by xundeenergie** [friendica/friendica#7037]
### Expected behavior
As in other socialmedia-sites friendica should be able to post a picture from clipboard. Take a screenshot with some tool (gnome-screenshot for example or greenshot on windows) to clipboard, and then post it (upload it) to friendica without saving a file in between. Just ctrl+c -> ctrl+v
php and javascript shoud provide this behaviour.

Thank yout

**Comment by ghost** [comment]
Hello, is this issue still open? Thank you

**Comment by tobiasd** [comment]
Hey @chynoradsky - yes the issue is still open, waiting for someone to pick it up and implement the feature. Are you interested?

**Comment by lapo-luchini** [comment]
[Some reference from StackOverflow](https://stackoverflow.com/a/6338207/166524) to help the attempt:

```javascript
// window.addEventListener('paste', ... or
document.onpaste = function(event){
  var items = (event.clipboardData || event.originalEvent.clipboardData).items;
  console.log(JSON.stringify(items)); // will give you the mime types
  for (index in items) {
    var item = items[index];
    if (item.kind === 'file') {
      var blob = item.getAsFile();
      var reader = new FileReader();
      reader.onload = function(event){
        console.log(event.target.result)}; // data url!
      reader.readAsDataURL(blob);
    }
  }
}
```

**Comment by MrPetovan** [comment]
The tricky part is that we use a custom photo upload system so any data that is collected this way needs to be first uploaded to get a photo reference id. It currently a tedious process in `mod/photos.php`, function `photos_post()`.

**Comment by xundeenergie** [comment]
when i look to facebook... a picture is uploaded, processed, then the thumbnail gets downloaded and showed... then i can post the whole thing... 

i'm back again on friendica 2 years later...

**Comment by MrPetovan** [comment]
What do you mean to say?

**Comment by annando** [comment]
@xundeenergie yeah, the issue is still open. We improved a lot of things in the two years, but some things still are open.

**Comment by xundeenergie** [comment]
> [Some reference from StackOverflow](https://stackoverflow.com/a/6338207/166524) to help the attempt:
> 
> ```js
> // window.addEventListener('paste', ... or
> document.onpaste = function(event){
>   var items = (event.clipboardData || event.originalEvent.clipboardData).items;
>   console.log(JSON.stringify(items)); // will give you the mime types
>   for (index in items) {
>     var item = items[index];
>     if (item.kind === 'file') {
>       var blob = item.getAsFile();
>       var reader = new FileReader();
>       reader.onload = function(event){
>         console.log(event.target.result)}; // data url!
>       reader.readAsDataURL(blob);
>     }
>   }
> }
> ```



I played around, to understand, how all works.

I could add a drop-action in the linkPreview.js so that a drag&drop of an url on desktop also creates a preview in composer.

And i found, that this funktion is maybe the right place for drag&drop and copy&paste inserting an image.

The problem is, ’event.clipboardData || event.originalEvent.clipboardData’ is always empty. Even if pasting or dropping the "file:///"-Link works.

There must be something done with copying the os-clipboard to browsers clipboard, and/or a async funktion to make this work.

Then i think i can build a funktion similar to githubs image-upload.

But now i need some hints to get the clipboard fetched into the browser.

**Comment by MrPetovan** [comment]
You need to use the result and feed it to an Ajax call to upload the file which should return a BBCode tag that you can then add to the active post/comment textarea. This calls may exist in the filebrowser but I can't confirm for sure right now.

**Comment by xundeenergie** [comment]
The BIG problem seems, that firefox can not use os clipboard (easy) with jquery. Firefox misses event.clipboardData...

I don't know, how drag&drop or copy&paste is realised in other projects, where all this works definitly...

There is a new inzerface via "navigator", but it exists only in the last firefox, not esr, which i use.

There is a project named "dropzone", which is available via composer too... but this could have potential to replace the whole upload-process including generation of previews...

**Comment by HankG** [comment]
When I added this capability to Diaspora I used the Fine Uploader JavaScript code which worked across all the browsers. The project isn't maintained any longer so I'm not sure it would be good to do that if I were to do it today but it may give hints about how they got these working on the different platforms. There may be some new equivalent too. Looking at how Mastodon web does it could provide some hints as well. https://fineuploader.com/

**Comment by xundeenergie** [comment]
> When I added this capability to Diaspora I used the Fine Uploader JavaScript code which worked across all the browsers. The project isn't maintained any longer so I'm not sure it would be good to do that if I were to do it today but it may give hints about how they got these working on the different platforms. There may be some new equivalent too. Looking at how Mastodon web does it could provide some hints as well. https://fineuploader.com/

Thanks.
I found some code in movim. It is obviously a kind of "where is the event listening" and of async request.

I try ro combine it with dropzone, which gives me previews...

I'm on learning, how this all works together.

**Comment by xundeenergie** [comment]
@MrPetovan my Pullrequest for easy-image-attach fixes this issue.
I don't know, how to connect this.

**Comment by MrPetovan** [comment]
Just write `Fixes #7037` in the description of your pull request, and it's going to be linked automatically.
